### PR TITLE
Hydjet++ 2.4.3 integration

### PR DESCRIPTION
**Pull request by wouf**
#### PR description:

Hydjet2 (HYDJET++) v2.4.3 integration as external generator.
Documentation for HYDJET++: http://lokhtin.web.cern.ch/lokhtin/hydjet++/documentation/
The minor changes in the PYQUEN and HYDJET v. 1.9.1 interfaces.

#### PR validation:

1) Hydjet2 - wf 160  with cms-sw/cmsdist#7483
2) Hydjet - wf 150
3) Pyquen - wf 300

### CODE REVIEW

**Comment by wouf** [comment]
Why does it deleted std::move from ref-qualified functions?

**Comment by srimanob** [comment]
test parameters:
- workflows = 160
- pull_request = https://github.com/cms-sw/cmsdist/pull/7483

**Comment by srimanob** [comment]
please test

**Comment by wouf** [comment]
Does HepPDT::ParticleDataTable deprecated? What have to be used instead?

**Comment by SiewYan** [comment]
Hello @wouf , I changes ```ESHandle<ParticleDataTable>``` to ```esConsumes<HepPDT::ParticleDataTable, PDTRecord>()```, the ESConsume is used to improve threading efficiency, for one of my PR [1]. This is in line with Issue [2]. Could you try to implement that? Thanks

[1] https://github.com/cms-sw/cmssw/pull/36125/files#diff-e98a0e405e813593a3117b0e41572a4803c0174aef5440a8d50e4b7618765730R98
[2] https://github.com/cms-sw/cmssw/issues/31061

**Comment by wouf** [comment]
@SiewYan I see, thank You!

**Comment by agrohsje** [comment]
Hi @wouf . Please check the log and fix so tests will proceed:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cae5ac/20993/headers_chks.log
Thanks.

**Comment by wouf** [comment]
@agrohsje Thanks, done.

File: GeneratorInterface/Hydjet2Interface/test/Hydjet2Analyzer.cc [hunk 1]
```diff
@@ -160,7 +160,9 @@ class Hydjet2Analyzer : public edm::EDAnalyzer {
   edm::InputTag genParticleSrc_;
   edm::InputTag genHIsrc_;
   edm::InputTag simVerticesTag_;
-  edm::ESHandle<ParticleDataTable> pdt;
+  //edm::ESHandle<ParticleDataTable> pdt;
```
Comment (srimanob) [inline GeneratorInterface/Hydjet2Interface/test/Hydjet2Analyzer.cc:163]: I think we should remove the ESHandle line forever. GIT will keep history anyways, no need to keep it in the code.

File: GeneratorInterface/HydjetInterface/test/HydjetAnalyzer.cc [hunk 2]
```diff
@@ -154,13 +154,15 @@ class HydjetAnalyzer : public edm::EDAnalyzer {
...
   edm::EDGetTokenT<edm::HepMCProduct> srcT_;
-  edm::EDGetTokenT<CrossingFrame<edm::HepMCProduct> > srcTmix_;
+  edm::EDGetTokenT<CrossingFrame<edm::HepMCProduct>> srcTmix_;
 
   // edm::InputTag src_;
   edm::InputTag genParticleSrc_;
   edm::InputTag genHIsrc_;
   edm::InputTag simVerticesTag_;
-  edm::ESHandle<ParticleDataTable> pdt;
+  //edm::ESHandle<ParticleDataTable> pdt;
```
Comment (srimanob) [inline GeneratorInterface/HydjetInterface/test/HydjetAnalyzer.cc:163]: same comment here.

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by srimanob** [comment]
+Upgrade

From the upgrade side, the PR introduces a new Hydjet2 workflow, 160.0_Hydjet2Q_MinBias_5020GeV. The new workflow runs fine.

File: GeneratorInterface/Hydjet2Interface/interface/Hydjet2Hadronizer.h [hunk 3]
```diff
@@ -1,49 +1,35 @@
...
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/Framework/interface/EDProducer.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include "SimDataFormats/CrossingFrame/interface/MixCollection.h"
+#include "SimDataFormats/PileupSummaryInfo/interface/PileupMixingContent.h"
+
+//#include "CLHEP/Random/RandomEngine.h"
+//#include "CLHEP/Random/RandFlat.h"
+//#include "CLHEP/Random/RandPoisson.h"
+//#include "CLHEP/Random/RandGauss.h"
```
Comment (agrohsje) [inline GeneratorInterface/Hydjet2Interface/interface/Hydjet2Hadronizer.h:23]: a long the lines of Phats comment: wouldn't it make sense to remove the CLHEP part as well?

File: GeneratorInterface/Hydjet2Interface/python/hydjet2DefaultParameters_cff.py [hunk 4]
```diff
@@ -2,76 +2,133 @@
 
 from Configuration.Generator.PythiaUESettings_cfi import *
 
+collisionParametersRHIC200GeV = cms.PSet(
+        fAw             = cms.double(197.0),    # beam/target atomic number
+        fSqrtS          = cms.double(200.0),   #
+	fMuB 		= cms.double(0.0285), 	# Chemical baryon potential per unit charge, fMuB [GeV]
+	fMuS 		= cms.double(0.007), 	# Chemical strangeness potential per unit charge, fMuS [GeV]
```
Comment (agrohsje) [inline GeneratorInterface/Hydjet2Interface/python/hydjet2DefaultParameters_cff.py:9]: The format is a bit screwed.

**Comment by agrohsje** [comment]
Thanks @wouf for the PR. Changes look good to me. Didn't spot anything crucial. I agree with Phat that one might fully remove the old parts that are commented out. Otherwise the format is sometimes screwed. Maybe worth fixing? 
@qliphy @perrotta : There are additional errors and warnings which show up in the comparison plots for Pyquen: 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2021-12-07-2300+cae5ac/47338/312.0_Pyquen_ZeemumuJets_pt10_2760GeV_2021+Pyquen_ZeemumuJets_pt10_2760GeV_2021+DIGIHI2021MIX+RECOHI2021MIX+HARVESTHI2021PPRECO/MessageLogger_Errors.html
Are they in the log:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2021-12-07-2300+cae5ac/47338/validateJR/all_OldVSNew_PyquenZEEMMppRECOin2021wf312p0/all_OldVSNew_PyquenZEEMMppRECOin2021wf312p0.log
I like to understand if they are relevant or not.

File: GeneratorInterface/Hydjet2Interface/interface/Hydjet2Hadronizer.h (hunk 3 above)
Comment (wouf) [reply GeneratorInterface/Hydjet2Interface/interface/Hydjet2Hadronizer.h]: Thanks, fixed.

File: GeneratorInterface/Hydjet2Interface/python/hydjet2DefaultParameters_cff.py (hunk 4 above)
Comment (wouf) [reply GeneratorInterface/Hydjet2Interface/python/hydjet2DefaultParameters_cff.py]: fixed

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by smuzaffar** [comment]
+hold
@wouf ,  `InitialParams.h` and `Hydjet2.h` should be picked up from `hydjet2` externals include directory. Why we have a copy of these on cmssw? We do pick the `libhydjet2.so` from `hydjet2` external then why we are not using the header files from external area too?

**Comment by wouf** [comment]
Dear @smuzaffar , it looks like I misunderstood You, when asked about such possibly. Can we also to use the tables from the sare/ folder?

**Comment by smuzaffar** [comment]
I thought you wanted to do that for your local tests. In cmssw official releases we should not do this type of hacks.
What is the reason to not get this headers from hydjet2/include directory itself? If you need to patch these headers then we can add a patch in hydjet2.spec

**Comment by wouf** [comment]
@smuzaffar No reason, opposite it would be more preferred, I just didn't understood that it's possible. So I will change it. What about data installed to the share/ folder from the tarball as well? ( instead of data/ folder in cmssw)

**Comment by smuzaffar** [comment]
@wouf , the data files are available under hydjet2/data directory. How are you going to access these files? If you are thinging to access these via `FileInPath` then we can just add the following in the hydjet2.xml 
```
<runtime name="CMSSW_SEARCH_PATH" default="$HYDJET2_BASE/data" type="path"/>
```

**Comment by wouf** [comment]
@smuzaffar Hydjet 2.4.3 used standard GNUInstallDirs during installation. Please try to build it again (it was switched off for cmssw, now I changed) and You can find folders: include, lib/lib64, share. All tables placing to share/. Could we use it from here?

**Comment by smuzaffar** [comment]
@wouf , please do not change the contents of tar file in place. `cmsBuild` is not going to download it as it has already cached the previous verison of tar file. I would suggest to move the code to github.
 
again, how are we going to access these data file[a] from cmssw? One way to find data files is to use [FileInPath](https://github.com/cms-sw/cmssw/search?q=FileInPath) and for this to work , we need to update hydjet2.xml to have `<runtime name="CMSSW_SEARCH_PATH" default="$HYDJET2_BASE/data" type="path"/>`


[a]
```
>ls hydjet2/2.4.3-7c1bce4049b934f8cce7987cb53ef48a/share
particles.data  tabledecay.txt
```

**Comment by wouf** [comment]
> `<runtime name="CMSSW_SEARCH_PATH" default="$HYDJET2_BASE/data" type="path"/>`

Don't we need something like this for include folder?
Or
`
    <environment name="HYDJET2_BASE" default="@TOOL_ROOT@"/>
    <environment name="LIBDIR" default="$HYDJET2_BASE/lib"/>
    <runtime name="CMSSW_SEARCH_PATH" default="$HYDJET2_BASE/share" type="path"/>
`
would be enough?

**Comment by smuzaffar** [comment]
yes, we also need this for INCLUDE , something like the following should be enough
```
<tool name="hydjet2" version="@TOOL_VERSION@">
  <lib name="hydjet2"/>
  <client>
    <environment name="HYDJET2_BASE" default="@TOOL_ROOT@"/>
    <environment name="LIBDIR" default="$HYDJET2_BASE/lib64"/>
    <environment name="INCLUDE" default="$HYDJET2_BASE/include"/>
  </client>
  <runtime name="CMSSW_SEARCH_PATH" default="$HYDJET2_BASE/data" type="path"/>
  <use name="pyquen"/>
  <use name="pythia6"/>
  <use name="lhapdf"/>
  <use name="root"/>
</tool>
```

I would also suggest that after line https://github.com/cms-sw/cmsdist/pull/7483/files#diff-07d9d263d0cb4c1ca6e3c32fe41c19fc658a03f8f9cbcfebb6d30334c88f0cbaR21 add something like
```
mkdir -p %{i}/data/externals/hydjet2
mv %{i}/share/* %{i}/data/externals/hydjet2
```

This will allow `scram` to properly create symlink for `hydjet2/data` under `cmssw/external/arch/data` and then one can use `cms.FileInPath('externals/hydjet2/datafile.txt')`  to find/load these data files.

**Comment by wouf** [comment]
@smuzaffar Thanks a lot!

**Comment by wouf** [comment]
> -code-checks
> 
> Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36316/27343
> 
> ERROR: Build errors found during clang-tidy run.
> 
> ```
> GeneratorInterface/Hydjet2Interface/interface/Hydjet2Hadronizer.h:26:10: error: 'Hydjet2.h' file not found [clang-diagnostic-error]
> #include "Hydjet2.h"
>          ^
> Suppressed 990 warnings (989 in non-user code, 1 with check filters).
> --
> GeneratorInterface/Hydjet2Interface/interface/Hydjet2Hadronizer.h:26:10: error: 'Hydjet2.h' file not found [clang-diagnostic-error]
> #include "Hydjet2.h"
>          ^
> Suppressed 1066 warnings (1065 in non-user code, 1 with check filters).
> --
> gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
> gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
> ```

@smuzaffar Does this happened due to the cash version of tarball have no headers? Do we have to iterate the version number for CMSSW?

**Comment by smuzaffar** [comment]
code-checks with cms.week1.PR_edda1a9e/52.0-b2970cfec92d880be7e91d763f01e9dc

**Comment by smuzaffar** [comment]
please test with cms-sw/cmsdist#7483

**Comment by smuzaffar** [comment]
test parameters:
- addpkg = Fireworks
- pull_requests = cms-sw/cmsdist#7483

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
please test

**Comment by srimanob** [comment]
Hi @smuzaffar @wouf 
Should we trigger the test with standard workflow 160 ? It is Hydjet2 which is the purpose of this PR, right? Thanks.

**Comment by smuzaffar** [comment]
sure @srimanob , I will update the test parameters and restart the test

**Comment by smuzaffar** [comment]
please test

**Comment by wouf** [comment]
Dear @smuzaffar , could You please to explain what is going on with this tests? Why it's failed?

**Comment by wouf** [comment]
@makortel what concurrent lumis is?

**Comment by makortel** [comment]
Framework's ability to process Events from multiple LuminosityBlocks concurrently, instead of synchronizing at LuminosityBlock boundaries. It was enabled by default for multithreaded jobs in cmsDriver in 11_3_0_pre1 (#32406), and at the framework level in 12_0_0_pre4 (#34231). For history of the migration of modules (and workarounds), see #25090.

**Comment by wouf** [comment]
> Framework's ability to process Events from multiple LuminosityBlocks concurrently, instead of synchronizing at LuminosityBlock boundaries. It was enabled by default for multithreaded jobs in cmsDriver in 11_3_0_pre1 (#32406), and at the framework level in 12_0_0_pre4 (#34231). For history of the migration of modules (and workarounds), see #25090.

Thank You!

**Comment by wouf** [comment]
@smuzaffar Cuold You please to run through these tests again?

**Comment by wouf** [comment]
@smuzaffar Could You please to run checks again?

**Comment by smuzaffar** [comment]
@wouf , I have started the tested for https://github.com/cms-sw/cmsdist/pull/7483, once the new hyjet external is deployed on cvmfs then I will start the code checks

**Comment by wouf** [comment]
> @wouf , I have started the tested for [cms-sw/cmsdist#7483](https://github.com/cms-sw/cmsdist/pull/7483), once the new hyjet external is deployed on cvmfs then I will start the code checks

I see, thanks!

**Comment by smuzaffar** [comment]
code-checks with cms.week1.PR_adfd7ee2/52.0-c70bc30a6458193435f279bf7f321f20

**Comment by smuzaffar** [comment]
please test

**Comment by srimanob** [comment]
Hi @wouf @smuzaffar 
Does this PR need to wait or to retest? Thanks.

**Comment by smuzaffar** [comment]
please test
@srimanob , last test was run more than a week ago, so lets refresh the tests.

**Comment by wouf** [comment]
@srimanob @smuzaffar  I don't really understand what's been going on here for the last few months. But checking cmsdist#7483 failed again...

**Comment by smuzaffar** [comment]
cms-sw/cmsdist#7483 only failed for `aarch64` ( https://github.com/cms-sw/cmsdist/pull/7483#issuecomment-1014845951 ) but that has nothing to do with this update. `aarch64` IBs also has the same failures.

**Comment by wouf** [comment]
Why we test it with RelVal 4.76? The only 160 (which is in this PR) is relevant, I guess

**Comment by SiewYan** [comment]
@wouf , the wf 4.76 test's failure is related to https://github.com/cms-sw/cmssw/issues/36771, and it is not related to this PR

**Comment by SiewYan** [comment]
@wouf , the wf 4.76 test's failure is related to https://github.com/cms-sw/cmssw/issues/36771, and it is not related to this PR.

**Comment by wouf** [comment]
> @wouf , the wf 4.76 test's failure is related to #36771, and it is not related to this PR

I see, thanks!

**Comment by srimanob** [comment]
+Upgrade

From the upgrade side, the change in `Configuration/PyReleaseValidation` is good.

**Comment by SiewYan** [comment]
from GEN side it is good too.

+Generators

**Comment by smuzaffar** [comment]
please test
relval-input test should pass now

**Comment by qliphy** [comment]
please test

**Comment by makortel** [comment]
The crash in 140.56 looks like #36804 that is being fixed in #36819

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by srimanob** [comment]
Kindly ping @cms-sw/pdmv-l2

**Comment by kskovpen** [comment]
+1

**Review by perrotta — COMMENTED** [review]
Please have also a look at the "Memory Leak" issues reported by the static analyzer
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cae5ac/22136/llvm-analysis/
at least for the two files that you are touching here: Hydjet2Hadronizer.cc and PyquenHadronizer.cc

File: GeneratorInterface/Hydjet2Interface/src/Hydjet2Hadronizer.cc [hunk 5]
```diff
@@ -63,37 +23,26 @@
...
 using namespace edm;
 using namespace std;
 using namespace gen;
 
-TString RunInputHYDJETstr;
-
-// definition of the static member fLastIndex
-int Particle::fLastIndex;
 bool ev = false;
+bool separateHydjetComponents_;
```
[the remark this answers is not in the document]
Comment (wouf) [reply GeneratorInterface/Hydjet2Interface/src/Hydjet2Hadronizer.cc:44]: I agree. But is it really matter? This bug is marked as Thread Safety, but Hydjet2 is not multi-threading itself due other reasons. I will fix it, but is it should be doing during this PR? I'm afraid, that with new fixes this PR will stuck few month more...

**Comment by wouf** [comment]
> Please have also a look at the "Memory Leak" issues reported by the static analyzer https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cae5ac/22136/llvm-analysis/ at least for the two files that you are touching here: Hydjet2Hadronizer.cc and PyquenHadronizer.cc

I will fix. Is it possible to execute static analyzer localy?

**Comment by wouf** [comment]
I can make changes to this PR, but I can't check them due to compilation issue:
`
$ scram b
********** ERROR: Missing Release top ************
  The release area "/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_12_2_X_2021-12-01-1100"
  for "CMSSW" version "CMSSW_12_2_X_2021-12-01-1100" is not available/usable.
  In case this release has been deprecated, you can move your code to
  one of the following release(s) of release series "CMSSW_12_2".

  CMSSW_12_2_0_pre1
  CMSSW_12_2_0_pre2
  CMSSW_12_2_0_pre3
  CMSSW_12_2_0
  CMSSW_12_2_0_patch1
  CMSSW_12_2_X_2022-01-30-0000
  CMSSW_12_2_X_2022-01-30-2300
  CMSSW_12_2_X_2022-01-31-1100
  CMSSW_12_2_X_2022-01-31-2300
  CMSSW_12_2_X_2022-02-01-1100
  CMSSW_12_2_X_2022-02-01-2300
  CMSSW_12_2_X_2022-02-02-1100
  CMSSW_12_2_X_2022-02-02-2300
  CMSSW_12_2_X_2022-02-03-1100
  CMSSW_12_2_X_2022-02-03-2300
  CMSSW_12_2_X_2022-02-04-1100
  CMSSW_12_2_X_2022-02-04-2300
  CMSSW_12_2_X_2022-02-05-1100
  CMSSW_12_2_X_2022-02-06-0000
  CMSSW_12_2_X_2022-02-06-2300
  CMSSW_12_2_X_2022-02-07-1100
  CMSSW_12_2_X_2022-02-07-2300
***********************************************
.SCRAM/slc7_amd64_gcc900/MakeData/src.mk:41: *** Release area has been removed/modified as /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_12_2_X_2021-12-01-1100/.SCRAM/slc7_amd64_gcc900/MakeData/DirCache.mk is missing..  Stop.
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
`
@smuzaffar is it possible to do something with it?

**Comment by wouf** [comment]
This memory leak was there many many years (for HYDJET, at least), maybe we can merge this PR and just after that I will create the new PR with fixes?

**Comment by smuzaffar** [comment]
@wouf , `CMSSW_12_2_X_2021-12-01-1100` IB is very old. We only need max two weeks old IBs. Please create a new cmssw dev area and move your "src" directory from existing dev area in to new one and rebuild.

**Comment by perrotta** [comment]
@wouf I understand this is a complicate PR, and I would like to avoid you mess it up if you rebase it on a more recent IB, as correctly suggested [above](https://github.com/cms-sw/cmssw/pull/36316#issuecomment-1032368819)

I think that all issues reported by the static analyzer are relevant, and witness some potentially dangerous inaccuracy in the code as it is implemented now. Some were pre-existent, and some were also introduced by this PR (e.g. `separateHydjetComponents_` potentially used uninitialized).

If you commit, as you say, in preparing quickly a patch with the corrections as soon as an integration build which contains this PR becomes available, we can even merge this one as it is now, and wait for the fixes. Please notice that pre5 must be cut in one week from now: let us know if you are able to provide the fix in a couple of days, so that it can be evaluated, signed and merged by then.

**Comment by perrotta** [comment]
+operations
- Assuming @wouf will fix the [remaining issues](https://github.com/cms-sw/cmssw/pull/36316#pullrequestreview-874923418) in a separate PR, once this one is merged

**Comment by perrotta** [comment]
+1

**Comment by wouf** [comment]
> 

I can suggest it right now (for memory leak we have just to free dynamicaly allocated memory). Could You please to explain how to commit such patch?

**Comment by perrotta** [comment]
> > 
> 
> I can suggest it right now (for memory leak we have just to free dynamicaly allocated memory). Could You please to explain how to commit such patch?

@wouf this PR is now merged. Just wait for the first IB which includes it (you can keep an eye on https://cmssdt.cern.ch/SDT/html/cmssdt-ib/#/ib/CMSSW_12_3_X for it) and implement your fixes based on that IB. Thank you!

**Comment by wouf** [comment]
> 

I see, thanks!

**Comment by wouf** [comment]
@SiewYan is something from cms-sw/cmssw#36574 should be added to the patch?

**Comment by SiewYan** [comment]
Hello @wouf , thanks for the reminder. There are efforts initiated from:
- #36404
- #31061 
to locate and migrate legacy EDModules to thread-efficient alternatives , as suggested in #36574

If the migration resume, will it break the integration effort as you mentioned in https://github.com/cms-sw/cmssw/pull/36574#issuecomment-1009681445?

**Comment by perrotta** [comment]
@wouf this PR is already available in CMSSW_12_3_X_2022-02-08-1100
By the way, in that IB there is a crash in wf 148.0 (HydjetQ_MinBias_XeXe_5442GeV_2017) which is quite likely related to this PR, and it should be also fixed:
```
----- Begin Fatal Exception 08-Feb-2022 [15:](https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc11/CMSSW_12_3_X_2022-02-08-1100/pyRelValMatrixLogs/run/148.0_HydjetQ_MinBias_XeXe_5442GeV_2017+HydjetQ_MinBias_XeXe_5442GeV_2017+DIGIHI2017+RECOHI2017+HARVESTHI2017/step1_HydjetQ_MinBias_XeXe_5442GeV_2017+HydjetQ_MinBias_XeXe_5442GeV_2017+DIGIHI2017+RECOHI2017+HARVESTHI2017.log#/15)13:48 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=HydjetGeneratorFilter label='generator'
Exception Message:
ValueError type of embeddingMode is expected to be int but declared as bool
----- End Fatal Exception ----------
```

**Comment by wouf** [comment]
@perrotta I will correct, thanks! It's related [this](https://github.com/cms-sw/cmssw/blob/master/Configuration/Generator/python/Hydjet_Quenched_MinBias_XeXe_5442GeV_cfi.py) config. 

@SiewYan It's already merged in CMSSW_12_3_X_2022-02-08-1100. Does the changes from [Generator Interface/Hydjet2Interface/test/Hydjet2Analyzer.cc](https://github.com/cms-sw/cmssw/pull/36574/files#diff-1ed4c2ff3c1d732c27d714a505dd7219425847c64645f2f335a05ad24a1407f9) is relevant? Or maybe You can look threw Hydjet2Analyzer.cc, HydjetAnalyzer.cc and PyquenAnalyzer.cc from CMSSW_12_3_X_2022-02-08-1100? As it now, it gives warning:
```
>> Compiling edm plugin /afs/cern.ch/work/w/wouf/TMP/2git/fix/CMSSW_12_3_X_2022-02-08-1100/src/GeneratorInterface/PyquenInterface/test/PyquenAnalyzer.cc
In file included from /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_3_X_2022-02-08-1100/src/FWCore/ParameterSet/interface/ParameterSetDescriptionFiller.h:27,
                 from /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_3_X_2022-02-08-1100/src/FWCore/ParameterSet/interface/ParameterSetDescriptionFillerPluginFactory.h:25,
                 from /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_3_X_2022-02-08-1100/src/FWCore/Framework/interface/MakerMacros.h:6,
                 from /afs/cern.ch/work/w/wouf/TMP/2git/fix/CMSSW_12_3_X_2022-02-08-1100/src/GeneratorInterface/PyquenInterface/test/PyquenAnalyzer.cc:12:
/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_3_X_2022-02-08-1100/src/FWCore/ParameterSet/interface/ParameterSetDescriptionFillerBase.h:106:65: warning: 'EDAnalyzer' is deprecated [-Wdeprecated-declarations]
  106 |     static const std::string& extendedBaseType(EDAnalyzer const*) { return kExtendedBaseForEDAnalyzer; }
      |                                                                 ^
/cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_3_X_2022-02-08-1100/src/FWCore/ParameterSet/interface/ParameterSetDescriptionFillerBase.h:107:65: warning: 'EDProducer' is deprecated [-Wdeprecated-declarations]
  107 |     static const std::string& extendedBaseType(EDProducer const*) { return kExtendedBaseForEDProducer; }
      |                                                                 ^
```

**Comment by SiewYan** [comment]
Those changes are relevant to suppress those warning messages. Please have a look at https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedAnalysisEDAnalyzer#Legacy_EDAnalyzer.

**Comment by perrotta** [comment]
@wouf indeed there are still quite several other places in which `embeddingMode` is inserted in the configs as a `bool`:
https://cmssdt.cern.ch/lxr/search?v=CMSSW_12_3_X_2022-02-07-2300&_casesensitive=1&_string=embeddingMode&_filestring=.py&%21v=CMSSW_12_3_X_2022-02-08-2300

Please have a  look at all them, and fix whenever needed. That could even go in a dedicated PR only meant to fix it, if it will be easier and faster to you doing as such instead of combining with all other updates and cleanings discussed here above.

**Comment by wouf** [comment]
@perrotta I'm not sure that configs like [Configuration/Generator/python/Pyquen2013Settings_cff.py](https://cmssdt.cern.ch/lxr/source/Configuration/Generator/python/Pyquen2013Settings_cff.py) still needed. It's more reasonable to delete it if it does not used any more. How to check all dependencies?

**Comment by wouf** [comment]
> @wouf , those changes are relevant to suppress those warning messages. Is that possible to open a PR, implement it on top of this PR? I can do that if you needed help.

@SiewYan yes, could You please!

**Comment by perrotta** [comment]
> @perrotta I'm not sure that configs like [Configuration/Generator/python/Pyquen2013Settings_cff.py](https://cmssdt.cern.ch/lxr/source/Configuration/Generator/python/Pyquen2013Settings_cff.py) still needed. It's more reasonable to delete it if it does not used any more. How to check all dependencies?

This should be agreed together with @cms-sw/generators-l2 
If they can be cleaned up, that's welcome,
In any case, if the configs remain in the release, they should better get fixed.

**Comment by wouf** [comment]
@perrotta @cms-sw/generators-l2 I'm afraid this question would be redirected to me (as HI GEN contact). Anyway files
[Configuration/Generator/python/Pyquen2013Settings_cff.py](https://cmssdt.cern.ch/lxr/source/Configuration/Generator/python/Pyquen2013Settings_cff.py) and  [Configuration/Generator/python/hydjet2DefaultParameters2015_cff.py](https://cmssdt.cern.ch/lxr/source/Configuration/Generator/python/hydjet2DefaultParameters2015_cff.py) does not used in any  Configuration/Generator cfi config.

**Comment by perrotta** [comment]
> @perrotta One more question: I can run llvm-analyser:
> 
> ```
> scram build -f compile COMPILER=llvm-analyzer SCRAM_IGNORE_MISSING_COMPILERS=yes
> ```
> 
> but it shows only memory leak bugs. How to see all types?

No idea, maybe @smuzaffar can help
I would personally trust what's reported by the static analyzer in these PR tests, or in the IBs:
https://cmssdt.cern.ch/SDT/html/cmssdt-ib/#/ib/CMSSW_12_3_X

**Comment by smuzaffar** [comment]
@wouf , as @perrotta mentioned, llvm-analyzer reports are available on the IB page. If you run locally then you can also see the same reports in your cmssw dev directory e..g
```
scram p CMSSW_12_3_X_2022-02-08-2300
cd CMSSW_12_3_X_2022-02-08-2300/
cmsenv
git cms-addpkg GeneratorInterface/Hydjet2Interface
scram build compile COMPILER=llvm-analyzer SCRAM_IGNORE_MISSING_COMPILERS=yes
```

and reports should be available in your dev area e.g. html file under `Hydjet2Hadronizer.plist`

**Comment by wouf** [comment]
> @wouf , as @perrotta mentioned, llvm-analyzer reports are available on the IB page. If you run locally then you can also see the same reports in your cmssw dev directory e..g
> 
> ```
> scram p CMSSW_12_3_X_2022-02-08-2300
> cd CMSSW_12_3_X_2022-02-08-2300/
> cmsenv
> git cms-addpkg GeneratorInterface/Hydjet2Interface
> scram build compile COMPILER=llvm-analyzer SCRAM_IGNORE_MISSING_COMPILERS=yes
> ```
> 
> and reports should be available in your dev area e.g. html file under `Hydjet2Hadronizer.plist`

Thank You @smuzaffar ! But there is only Memory Error, how to see also ThreadSafety, like [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cae5ac/22136/llvm-analysis/)?

**Comment by smuzaffar** [comment]
I think for these static checks you need to do something like https://github.com/cms-sw/cms-bot/blob/master/pr_testing/test_multiple_prs.sh#L773-L776 e.g.
```
git cms-addpkg Utilities/StaticAnalyzers
export USER_CXXFLAGS='-Wno-register -DEDM_ML_DEBUG -w'
export SCRAM_IGNORE_PACKAGES='Fireworks/% Utilities/StaticAnalyzers'
export USER_LLVM_CHECKERS='-enable-checker threadsafety -enable-checker cms -enable-checker deprecated -disable-checker cms.FunctionDumper'
scram b -k -j 32 checker SCRAM_IGNORE_SUBDIRS=test
```

**Comment by wouf** [comment]
> I think for these static checks you need to do something like https://github.com/cms-sw/cms-bot/blob/master/pr_testing/test_multiple_prs.sh#L773-L776 e.g.
> 
> ```
> git cms-addpkg Utilities/StaticAnalyzers
> export USER_CXXFLAGS='-Wno-register -DEDM_ML_DEBUG -w'
> export SCRAM_IGNORE_PACKAGES='Fireworks/% Utilities/StaticAnalyzers'
> export USER_LLVM_CHECKERS='-enable-checker threadsafety -enable-checker cms -enable-checker deprecated -disable-checker cms.FunctionDumper'
> scram b -k -j 32 checker SCRAM_IGNORE_SUBDIRS=test
> ```

Thank You!

**Comment by wouf** [comment]
> > @wouf , those changes are relevant to suppress those warning messages. Is that possible to open a PR, implement it on top of this PR? I can do that if you needed help.
> 
> @SiewYan yes, could You please!

@SiewYan Do You already created such PR? If not I can try to include it to the memory leak bug fix PR, I'm preparing right now.

**Comment by SiewYan** [comment]
@wouf , I have not, you may proceed with the fix in the PR. thank you.

**Pull request merged**